pkg/controller/controller_utils.go

```diff
@@ -271,21 +269,28 @@ func filterActivePods(pods []api.Pod) []*api.Pod {
 	return result
 }
 
-// updateReplicaCount attempts to update the Status.Replicas of the given controller, with retries.
-// Note that the controller pointer might contain a more recent version of the same controller passed into the function.
-func updateReplicaCount(rcClient client.ReplicationControllerInterface, controller *api.ReplicationController, numReplicas int) error {
-	return wait.Poll(10*time.Millisecond, 100*time.Millisecond, func() (bool, error) {
-		if controller.Status.Replicas != numReplicas {
-			glog.V(4).Infof("Updating replica count for rc: %v, %d->%d", controller.Name, controller.Status.Replicas, numReplicas)
-			controller.Status.Replicas = numReplicas
-			_, err := rcClient.Update(controller)
-			if err != nil {
-				glog.V(2).Infof("Controller %v failed to update replica count: %v", controller.Name, err)
-				// Update the controller with the latest resource version for the next poll
-				controller, _ = rcClient.Get(controller.Name)
-				return false, err
-			}
+// updateReplicaCount attempts to update the Status.Replicas of the given controller, with a single GET/PUT retry.
+func updateReplicaCount(rcClient client.ReplicationControllerInterface, controller api.ReplicationController, numReplicas int) (updateErr error) {
+	// This is the steady state. It happens when the rc doesn't have any expectations, since
+	// we do a periodic relist every 30s.
+	if controller.Status.Replicas == numReplicas {
+		return nil
+	}
+	var getErr error
+	glog.V(4).Infof("Updating replica count for rc: %v, %d->%d", controller.Name, controller.Status.Replicas, numReplicas)
+	for i, rc := 0, &controller; i < 2; i++ {
+		rc.Status.Replicas = numReplicas
+		_, updateErr = rcClient.Update(rc)
+		if updateErr == nil {
+			return nil
+		}
+		// Update the controller with the latest resource version for the next poll
+		if rc, getErr = rcClient.Get(controller.Name); getErr != nil {
```

Comment:
This will result in an extra get if the update fails twice but I prefer the loop since it gives me room to increase retries and is clearer than the unrolled version. I also want to allow the first update to happen naturally since there's a good chance it'll pass. Thoughts?